api/envoy/service/ext_proc/v3alpha/external_processor.proto

```diff
@@ -0,0 +1,273 @@
+    TrailersResponse response_trailers = 6;
+
+    // If specified, return the specified response downstream to the client
+    // and end further filter processing on the request path.
+    ImmediateResponse immediate_response = 7;
+  }
+
+  // Add new metadata to the Envoy context
+  google.protobuf.Struct new_metadata = 8;
+
+  // Override how parts of the HTTP request and response are processed
+  // for the duration of this particular request/response only. Servers
+  // may use this to intelligently control how requests are processed
+  // based on the headers and other metadata that they see.
+  envoy.extensions.filters.http.ext_proc.v3alpha.ProcessingMode mode_override = 9;
+
+  // Change how and if the filter waits for responses from the server
+  // for future messages. Servers may use this to switch between
+  // synchronous and asynchronous processing modes based on what they see.
+  // A message with this field may be sent even if none of the "oneof" fields
+  // have been sent, and it may be sent out-of-band with respect to
+  // the other messages if desired.
+  ProcessingModeChange mode_change = 10;
+}
+
+// The following are messages that are sent to the server.
+
+// This message is sent to the external server when the HTTP request and responses
+// are first received.
+message HttpHeaders {
+  // The HTTP request headers. All header keys will be
+  // lower-cased, because HTTP header keys are case-insensitive.
+  config.core.v3.HeaderMap headers = 1;
+
+  // Information about the client's connection -- downstream if this is
+  // a request and upstream if it is a response.
+  auth.v3.AttributeContext.Peer client_info = 2;
+
+  // The values of properties selected by the "request_properties"
+  // list in the configuration. Each entry in the map is the
+  // value of the specified property.
+  map<string, google.protobuf.Struct> properties = 3;
+
+  // If true, then there is no message body associated with this
+  // request or response.
+  bool end_of_stream = 4;
+}
+
+// This message contains the message body that Envoy sends to the external server.
+message HttpBody {
+  bytes body = 1;
+
+  bool end_of_stream = 2;
+}
+
+// This message contains the trailers.
+message HttpTrailers {
+  config.core.v3.HeaderMap trailers = 1;
+}
+
+// The following are messages that may be sent back by the server.
+
+// This message must be sent in response to a request_headers or response_headers
+// message.
+message HeadersResponse {
+  // Instructions on how to manipulate the request headers
+  HeaderMutation header_mutation = 1;
+
+  // If specified, add a body to the message if none was present
+  // before.
+  BodyMutation body_mutation = 2;
+}
+
+// This message must be sent in response to a request_trailers
+// or response_trailers message.
+message TrailersResponse {
+  // Instructions on how to manipulate the request trailers
+  HeaderMutation header_mutation = 1;
+}
+
+message BodyResponse {
```

Comment:
Same comment about trailers and end_stream

Reply:
See above, essentially we are talking about the ability to end a stream early, or *not* end a stream as the case may be.